### DISCUSSION TITLE
can't find sodium.h

### DISCUSSION THREAD

**Discussion by hzkincony** [HomeSpan/HomeSpan#80] [General]
hello. When i install library online  and compile in arduino 1.8.13, it have follows:
C:\Users\xxx\Documents\Arduino\libraries\HomeSpan\src\HAP.cpp:30:20: fatal error: sodium.h: No such file or directory
i can't find the sodium.h file, where is it? Thanks.

**Comment by HomeSpan** [comment]
The sodium file, as well as numerous others, should be part of the Arduino-ESP32 board definitions.  Have you installed Arduino-ESP32?  Also, when compiling, have you selected an ESP32 board (as opposed to an Arduino board)?

**Comment by hzkincony** [comment]
thank you, i have solved the problem, re install esp32 by online.
another question: i want to know, how to connect DI button for 4 lamp output, i have see the sample-15 button for demo. But i don't understand, how to define the button PIN for ESP32, i just want use 4 normal-open switch button to turn ON/OFF lamp (relay) and the relay state will update with homekit app. many thanks for you have support this good plugin.

**Reply by zIlgar** [reply]
I have same issue right now. what should I do to make it work. also I don't understand what you reinstalled?

**Reply by HomeSpan** [reply]
Note HomeSpan will only compile under version 2.X of the Arduino-ESP32 board manager.  Not compatible with version 3.X.

**Reply by zIlgar** [reply]
> Note HomeSpan will only compile under version 2.X of the Arduino-ESP32 board manager.  Not compatible with version 3.X. 

Thanks very much it worked🙏

**Comment by HomeSpan** [comment]
You will want to use two separate pins.  One pin should be set to control the relay to turn on the lamp.  This pin is controlled by HomeKit and will enable you to turn on the lamp from the Home App.  You would then use a second pin to install a push button (connecting the pin you choose to ground when pressed).  This is the pin on which you create a SpanButton() object.  You then include logic in the lamp service (LightBulb Service) based on the examples that looks for button presses, and when found it changes the output of the pin that is connected to the relay, as well as sends HomeKit a notification that the lamp status has changed.

**Comment by hzkincony** [comment]
i have used "05-WorkingLED" sample, it work well. do you have a full sample that for control relay , where can see define the relay I/O pin and switch button I/O pin. I have see your sample "15-RealPushButtons", that is for dimmer. I don't know how to use the sample, do you have more details or video tour? I just want to use 4 relays, can you help me, Thank you very much.

**Comment by RichiG88** [comment]
with the new Arduino-ESP32 3.0.0 board definitions, the libsodium library is not included anymore. I can't compile my homespan sketch with the new board definitions. Can this be fixed? Thanks in advance.

**Reply by HomeSpan** [reply]
I will have to check with the Arduino-Esp32 folks.

**Reply by Paffdsdaf** [reply]
Oh thank you I have same problem and searching how fix problem.

**Reply by harshal17** [reply]
thanks. Having the same issue. It can't find sodium.h in HAP, HomeSpan and SRP.cpp files

/Arduino/libraries/HomeSpan/src/HAP.cpp:32:10: fatal error: sodium.h: No such file or directory
   32 | #include <sodium.h>

**Reply by Glujaz** [reply]
Revert back to the previous espressos library. I use 2.0.17 and it works great (Homespan suggest 2.0.14)

**Reply by Paffdsdaf** [reply]
thank you its works
